dev/core#2800 Fix bounce processing to handle verp emails

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
dev/core#2800 Fix bounce processing to handle verp emails

https://lab.civicrm.org/dev/core/-/issues/2800

Before
----------------------------------------
This fixes 2 issues when combining bounce processing with verp emails

1) the emails being processed are being matched to email they come from - ie the verp email. Hence the created activities are not linked to the actual contact and instead create endless variants of the sender - e.g `b.123.6789.zombie@civicrm.org`. 

This happens with both the job.process_activities job and with job.fetch_bounces with is_create_activities = TRUE. 

2) this second one is a doozy & probably only showed up in tests because the tests mostly cover obscure email formats. The create activities code does a call to get the body and attachments.  importantly, it moves those attachments to the civicrm files as part of creating the activity. Then there is a second attempt to get the email body in the bounce processing. This one tried to load the full email and fails when it looks for the attachment in the location it has already been moved from. 


After
----------------------------------------
1) This fix adapts the handling to identify verp emails (using existing regex) and look up the contact ID from the mailing_event_queue. This is used as the source_contact_id (I didn't add a target or assignee at this stage cos I wasn't sure if that was just data-cruft or useful stuff - but it would be retrieved from the mailing)  If the activity can't be created the email is skipped - because we wouldn't be able to process it through to actions if we can't get a valid contact ID anyway.

2) The body is loaded once using the better of the 2 bits of code to load it & the same value is used in both places. Missing attachments don't cause crashes

Technical Details
----------------------------------------
This is against 5.66 to keep it in the same release as the preliminary cleanup work for this & to allow consolidated testing.

The code used to get the email body in fetch activities looks like it came from a tutorial on how to best use ezcmail. It catches all our test examples without falling back to 'last gasp code'.

The code used to getBody() for bounces looks like it was ... written by us. It does a subset of the processing in earlier call to get body and in many of our test cases falls back on last gasp code. This last gasp code is where it tries to load the attachments again - and falls over.

I dug into the 2 functions and I feel pretty sure the bounce one doesn't add any additional value but DOES have less capability than the other. So I removed that code in favour of the code that does the job much more thoroughly

Unfortunately the person received weird email from Exchange 2003 did not log the headers https://issues.civicrm.org/jira/browse/CRM-9361 - but even if the preferred version of the function doesn't parse that there is default text so it won't fail & it seems like the risk of having a bit less info about an (untested) obscure email bounce body is not that probematic

Comments
----------------------------------------
Good test cover in CRM_Utils_Mail_EmailProcessorTest

eileenmcnaughton committed:
Move from handling to processor, test

eileenmcnaughton committed:
dev/core#2800 Fix bounce processing to handle verp emails

This fixes 2 issues when combining bounce processing with verp emails

1) the emails being processed are being matched to email they come from -
ie the verp email. Hence the created activities are not linked to the
actual contact and instead create endless variants of the sender.
This happens with both the job.process_activities job and with
job.fetch_bounces with is_create_activities = TRUE.
This fix adapts the handling to identify verp emails (using existing regex)
and look up the contact ID from the mailing_event_queue. This is
used as the source (I didn't add a target or assignee at this stage cos
I wasn't sure if that was just data-cruft or useful stuff - but it would
be retrieved from the mailing)

2) this second one is a doozy & probably only showed up in tests because the
tests mostly cover obscure email formats. The create activities code does
a call to get the body and attachments using a function that looks like it
came from a tutorial on how to best use ezcmail. It seems to parse
all the email variants and attachments and, importantly, moves those attachments
to the civicrm files as part of creating the activity.

The code used to getBody() for bounces looks like it was ... written by us.
It does a subset of the processing in earlier call to get body and in
many of our test cases falls back on a blunt generateBody(). This call parses the
email, including the attachments, which are no longer there cos they got moved....
so it fatals.

I dug into the 2 functions and I feel pretty sure the bounce one doesn't add any
additional value but DOES have less capability than the other. So I removed that code,

Unfortunately the person received weird email from Exchange 2003 did not log
the headers https://issues.civicrm.org/jira/browse/CRM-9361 - but even if the
preferred version of the function doesn't parse that there is default text so
it won't fail & it seems like the risk of having a bit less info about an (untested)
obscure email bounce body is not that probematic

eileenmcnaughton committed:
Move handling of empty from obscure function to main email handler class

I wondered if the zeta components update would make this fix obsolete but adding
a test  says not...

So this locks in the fix on
https://issues.civicrm.org/jira/browse/CRM-19215

agileware-justin commented:
@eileenmcnaughton Agileware will have email examples and email headers for this issue - part of the investigation deep-dive performed for the CiviCRM support requests. Would that be helpful?

eileenmcnaughton commented:
@agileware-justin yes - PR them into this directory

https://github.com/civicrm/civicrm-core/tree/a32a65e3c9615e13f1f14d27e9c85f37097e0192/tests/phpunit/CRM/Utils/Mail/data

Note we have agreement to put these against the rc for the next day or 2 to try to keep this work in one release

eileenmcnaughton commented:
@larssandergreen @agileware-justin so I just deployed this into production. It worked 

However, it immediately highlighted an issue... We have 51 million bounces in our civicrm_mailing_event_bounce table. Had we been running this all along that would be an additional 51 million activites of type 'Inbound Email' in our civicrm_activity table. Importantly we would not be able to implement a retention policy because we would want to delete the details of a bounce after a few months, but not the details of a real reply.

I feel like we want an activity type of 'Bounce' and to use that if it exists for bounces when creating activities.

Perhaps for the rc I'll just add to check if it exists & use it if it does & we can look at adding it later for all installs as it all shakes down (we'll always need the if-exists check in case people delete/ disable)

eileenmcnaughton commented:
Hmm - I'm also feeling like I want to combine in some bounce type details into the body - so it's visible what decisions were made about putting the email on hold

eileenmcnaughton commented:
I just pushed in the commits from https://github.com/civicrm/civicrm-core/pull/27352 - that makes the changes a little harder to read but @larssandergreen has already had a go at that. However, it means that this PR + the rc tarball is what we have just deployed into production & so far is working well

eileenmcnaughton committed:
Use bounce activity type, if exists

We can look at adding in upgrade script in later version after bedding in

eileenmcnaughton committed:
Add detail about bounce to activity

Note adding fetch_activities to the test exposed a mysql error so I had
to add handling there

eileenmcnaughton commented:
test this please

colemanw commented:
Unit tests + `r-run` in prod is pretty solid.
